### PR TITLE
[FLINK-24607] Make OperatorCoordinator closing sequence more robust.

### PR DESCRIPTION
**Pull request by becketqin**
## What is the purpose of the change
Currently the closing sequence of the `OperatorCoordinator` does not swallow the exceptions thrown in the middle. So when exception is thrown in the middle, the rest of the closing sequence will be skipped. This patch makes the closing sequence of `OperatorCoordinator` more robust by swallowing the exception thrown in the middle, and also retries interrupting the threads in case the the closing sequence is blocked multiple times.

Some of the methods, such as `closeQuietly()`, in the `IOUtils` seem better to be put in `ComponentClosingUtils`. Will do that in a separate PR. 

## Brief change log
0dfe9b7 Add util methods to shutdown executor services.
1cb9f56 Make OperatorCoordinator closure more robust.

## Verifying this change
This change added the following unit tests:

1. `ComponentClosingUtilsTest` class.
2. `SourceCoordinatorTest.testBlockOnClose()` test method.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by becketqin** [comment]
Ping @wuchong @pnowojski for review.

**Comment by wuchong** [comment]
cc @leonardBang

**Comment by wuchong** [comment]
@becketqin , do you want this PR get merged in 1.15 release?

**Comment by becketqin** [comment]
@wuchong Yes, I think it should be in 1.15. This is a bug fix, so technically speaking, it should be OK to merge even after code freeze.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java [hunk 1]
```diff
@@ -217,18 +207,11 @@ public void start() throws Exception {
...
     public void close() throws Exception {
         LOG.info("Closing SourceCoordinator for source {}.", operatorName);
-        try {
-            if (started) {
-                context.close();
-                if (enumerator != null) {
-                    enumerator.close();
-                }
+        if (started) {
+            closeQuietly(context);
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:211]: Could we use `IOUtils.closeAll` instead? That way we would not loose exception from `context.close()`.

```
IOUtils.closeAll(context, enumerator);
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java [hunk 2]
```diff
@@ -82,7 +84,7 @@ Licensed to the Apache Software Foundation (ASF) under one
 
     private static final Logger LOG = LoggerFactory.getLogger(SourceCoordinatorContext.class);
 
-    private final ExecutorService coordinatorExecutor;
+    private final ScheduledExecutorService coordinatorExecutor;
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:87]: nit: Could we create the `coordinatorExecutor` in the ctor? It's a bit weird now that we close a `coordinatorExecutor` that we are given from the outside of the class. As far as I checked it's always a new service created in the `SourceCoordinatorProvider`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java [hunk 3]
```diff
@@ -259,9 +265,9 @@ public void runInCoordinatorThread(Runnable runnable) {
     @Override
     public void close() throws InterruptedException {
         closed = true;
-        notifier.close();
-        coordinatorExecutor.shutdown();
-        coordinatorExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        // Close quietly so the closing sequence will be executed completely.
+        closeQuietly(notifier);
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:269]: Can we use `IOUtils.closeAll` here as well?

```
        IOUtils.closeAll(
                notifier,
                () ->
                        shutdownExecutorForcefully(
                                coordinatorExecutor, Duration.ofMillis(Long.MAX_VALUE)));
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java [hunk 4]
```diff
@@ -153,7 +156,6 @@ public void close() throws InterruptedException {
             return;
         }
         // Shutdown the worker executor, so no more worker tasks can run.
-        workerExecutor.shutdownNow();
-        workerExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+        shutdownExecutorForcefully(workerExecutor, Duration.ofMillis(Long.MAX_VALUE));
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java:159]: nit: I find it equally weird, similarly as in `SourceCoordinatorContext`, that we close something, given from the outside.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java [hunk 5]
```diff
@@ -95,8 +100,94 @@ private ComponentClosingUtils() {}
...
+     * Shutdown the given executor forcefully within the given timeout.
+     *
+     * @param executor the executor to shut down.
+     * @param timeout the timeout duration.
+     * @param interruptable when set to true, the method cannot be interrupted. Each interruption to
+     *     the thread results in another {@code ExecutorService.shutdownNow()} call to the shutting
+     *     down executor.
+     * @return true if the given executor is terminated, false otherwise.
+     */
+    public static boolean shutdownExecutorForcefully(
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java:149]: Could we rewrite it using `Deadline`?

```
    public static boolean shutdownExecutorForcefully(
            ExecutorService executor, Duration timeout, boolean interruptable) {
        boolean isInterrupted = false;
        final Deadline deadline = Deadline.fromNowWithClock(timeout, clock);
        do {
            executor.shutdownNow();
            try {
                executor.awaitTermination(deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS);
            } catch (InterruptedException e) {
                isInterrupted = interruptable;
            }
        } while (!executor.isTerminated() && !isInterrupted && deadline.hasTimeLeft());
        return executor.isTerminated();
    }
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java [hunk 6]
```diff
@@ -95,8 +100,94 @@ private ComponentClosingUtils() {}
...
+        do {
+            executor.shutdownNow();
+            try {
+                long timeRemaining = timeout.toMillis() - timeElapsed;
+                executor.awaitTermination(timeRemaining, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                isInterrupted = interruptable;
+            }
+            timeElapsed = clock.relativeTimeMillis() - startingTime;
+        } while (!executor.isTerminated() && !isInterrupted && timeElapsed < timeout.toMillis());
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java:163]: nit: How about we change the order? I guess `executor.isTerminated()` is the most expensive operation and we do not need to execute it here if other conditions are not met.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java [hunk 7]
```diff
@@ -95,8 +100,94 @@ private ComponentClosingUtils() {}
...
+        return executor.isTerminated();
+    }
+
     static void abortThread(Thread t) {
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java:167]: Why is it not private?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/ManuallyTriggeredScheduledExecutorService.java [hunk 8]
```diff
@@ -113,7 +113,7 @@ public boolean isTerminated() {
     }
 
     @Override
-    public boolean awaitTermination(long timeout, TimeUnit unit) {
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
```
Comment (dawidwys) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/ManuallyTriggeredScheduledExecutorService.java:116]: That's an unnecessary change, isn't it?

**Review by becketqin — COMMENTED** [review]
@dawidwys Thanks for the review. I have addressed / replied to the comments. Would you take another look? Thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java (hunk 1 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:211]: I get that. Is it mandatory that we do not have any exceptions? Or just that if an exception is thrown in `context.close` we continue with `enumerator.close`. That's what `closeAll` does. The difference is though at the end it will throw the exception from the `context#close` and if both `context.close` and `enumerator.close` throw the second one will be thrown with the first as suppressed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java (hunk 3 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java]: By default we "suppress" all exceptions, because the default `suppressedException` is `Exception.class` and we suppress all exception that can be assigned to `Exception`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java [hunk 9]
```diff
@@ -95,8 +100,94 @@ private ComponentClosingUtils() {}
...
+            } catch (InterruptedException e) {
+                // Let it go.
+            }
+        }
+    }
+
+    // ========= Method visible for testing ========
+
+    @VisibleForTesting
+    static void setClock(Clock clock) {
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java:188]: My eyes hurt if I see a static variable being assigned, but I won't be stubborn on it ;) Up to you to decide.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java (hunk 6 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/ComponentClosingUtils.java]: Makes sense.

File: flink-core/src/main/java/org/apache/flink/api/common/time/Deadline.java [hunk 10]
```diff
@@ -103,11 +107,23 @@ public static Deadline fromNow(Duration duration) {
...
     }
 
     @Override
     public String toString() {
         return LocalDateTime.now().plus(timeLeft()).toString();
     }
+
+    // -------------------- private helper methods ----------------
+
+    private static long addHandlingOverflow(long x, long y) {
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/common/time/Deadline.java:121]: nit: we could add a comment it has been copied over from `Math#addExact` and we changed the overflow behaviour.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java (hunk 3 above)
Comment (becketqin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java]: I see. However, here I feel it is slightly more readable to just shutdown the two executors explicitly. It seems quite simple and straight forward compared with constructing two `AutoClosables` and invoke `closeAll()`.

**Review by dawidwys — APPROVED** [review]

**Comment by dawidwys** [comment]
Sure, makes sense. Thanks for updates. LGTM
